// @devexpress/analytics-core/query-builder/metadata.js
/**
* DevExpress Analytics (query-builder\metadata.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var editorTemplates_1 = require("./widgets/editorTemplates");
var size_1 = require("../core/elements/size");
var point_1 = require("../core/elements/point");
exports.name = { propertyName: 'name', modelName: '@Name', displayName: 'Name', localizationId: 'DevExpress.DataAccess.Sql.SqlQuery.Name', disabled: true, editor: editorTemplates_1.editorTemplates.getEditor('text') };
exports.alias = { propertyName: 'alias', modelName: '@Alias', displayName: 'Alias', localizationId: 'DataAccessUIStringId.QueryBuilderColumns_Alias', defaultVal: '', editor: editorTemplates_1.editorTemplates.getEditor('text') };
exports.text = { propertyName: 'text', modelName: '@Text', displayName: 'Text', editor: editorTemplates_1.editorTemplates.getEditor('text') };
exports.selected = { propertyName: 'selected', displayName: 'Output', editor: editorTemplates_1.editorTemplates.getEditor('bool'), localizationId: 'DataAccessUIStringId.QueryBuilderColumns_Output' };
exports.size = { propertyName: 'size', modelName: '@Size', defaultVal: '100,125', from: size_1.Size.fromString };
exports.location = { propertyName: 'location', modelName: '@Location', from: point_1.Point.fromString };
exports.sizeLocation = [exports.size, exports.location];
exports.unknownSerializationsInfo = [exports.name].concat(exports.sizeLocation);
