// @devexpress/analytics-core/widgets/criteria/operators/unary.js
/**
* DevExpress Analytics (widgets\criteria\operators\unary.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var criteriaOperator_1 = require("./criteriaOperator");
var UnaryOperatorType;
(function (UnaryOperatorType) {
    UnaryOperatorType[UnaryOperatorType["Minus"] = 0] = "Minus";
    UnaryOperatorType[UnaryOperatorType["Plus"] = 1] = "Plus";
    UnaryOperatorType[UnaryOperatorType["BitwiseNot"] = 2] = "BitwiseNot";
    UnaryOperatorType[UnaryOperatorType["Not"] = 3] = "Not";
    UnaryOperatorType[UnaryOperatorType["IsNull"] = 4] = "IsNull";
})(UnaryOperatorType = exports.UnaryOperatorType || (exports.UnaryOperatorType = {}));
var UnaryOperator = (function (_super) {
    __extends(UnaryOperator, _super);
    function UnaryOperator(operatorType, operand) {
        var _this = _super.call(this) || this;
        _this.type = 'Unary';
        _this.operand = operand || new criteriaOperator_1.CriteriaOperator();
        _this.operatorType = operatorType;
        return _this;
    }
    Object.defineProperty(UnaryOperator.prototype, "leftPart", {
        get: function () {
            return this.operand;
        },
        enumerable: true,
        configurable: true
    });
    UnaryOperator.prototype.assignFrom = function (criteriaOperator) {
        if (this.operatorType === UnaryOperatorType.Not) {
            if (criteriaOperator instanceof UnaryOperator) {
                this.operand.assignFrom(criteriaOperator.operand);
            }
            else {
                this.operand.assignFrom(criteriaOperator);
            }
        }
        else {
            if (criteriaOperator instanceof UnaryOperator) {
                this.operand = criteriaOperator.operand.leftPart;
            }
            else {
                this.operand = criteriaOperator.leftPart || criteriaOperator;
            }
        }
    };
    Object.defineProperty(UnaryOperator.prototype, "displayType", {
        get: function () {
            return UnaryOperatorType[this.operatorType];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UnaryOperator.prototype, "enumType", {
        get: function () {
            return UnaryOperatorType;
        },
        enumerable: true,
        configurable: true
    });
    UnaryOperator.prototype.accept = function (visitor) {
        return visitor.visitUnaryOperator
            ? visitor.visitUnaryOperator(this)
            : new UnaryOperator(this.operatorType, this.operand && this.operand.accept(visitor));
    };
    return UnaryOperator;
}(criteriaOperator_1.CriteriaOperator));
exports.UnaryOperator = UnaryOperator;
