// @devexpress/analytics-core/query-builder/widgets/masterdetaileditor/_masterDetailEditorPopupManager.js
/**
* DevExpress Analytics (query-builder\widgets\masterdetaileditor\_masterDetailEditorPopupManager.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MasterDetailEditorPopupManager = (function () {
    function MasterDetailEditorPopupManager(target, popupService, action, popupItems) {
        var _this = this;
        this.showPopup = function (_, element) {
            if (_this._popupService['subscription']) {
                _this._popupService['subscription'].dispose();
            }
            _this._popupService.title('');
            _this._updateActions(_this.target);
            _this._popupService.target(element);
            _this._popupService.visible(true);
        };
        this.target = target;
        this._action = action;
        this._popupService = popupService;
        this._popupItems = popupItems;
    }
    MasterDetailEditorPopupManager.prototype._updateActions = function (viewModel) {
        var _this = this;
        this._popupService['subscription'] = this._popupService.visible.subscribe(function (newVal) {
            _this.target.isSelected && _this.target.isSelected(newVal);
        });
        this._popupService.data({
            data: this._popupItems,
            template: 'dx-filtereditor-popup-common',
            click: function (data) {
                viewModel[_this._action](data);
                _this._popupService.visible(false);
            }
        });
    };
    return MasterDetailEditorPopupManager;
}());
exports.MasterDetailEditorPopupManager = MasterDetailEditorPopupManager;
