// devextreme/cjs/ui/grid_core/ui.grid_core.editing_constants.js
/**
 * DevExtreme (cjs/ui/grid_core/ui.grid_core.editing_constants.js)
 * Version: 21.2.7
 * Build date: Mon Apr 11 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.VIEWPORT_TOP_NEW_ROW_POSITION = exports.VIEWPORT_BOTTOM_NEW_ROW_POSITION = exports.TARGET_COMPONENT_NAME = exports.ROW_SELECTED_CLASS = exports.ROW_CLASS = exports.ROW_BASED_MODES = exports.PAGE_TOP_NEW_ROW_POSITION = exports.PAGE_BOTTOM_NEW_ROW_POSITION = exports.MODES_WITH_DELAYED_FOCUS = exports.LAST_NEW_ROW_POSITION = exports.FOCUSABLE_ELEMENT_SELECTOR = exports.FIRST_NEW_ROW_POSITION = exports.EDIT_MODE_ROW = exports.EDIT_MODE_POPUP = exports.EDIT_MODE_FORM = exports.EDIT_MODE_CELL = exports.EDIT_MODE_BATCH = exports.EDIT_MODES = exports.EDIT_FORM_CLASS = exports.EDITOR_CELL_CLASS = exports.EDITORS_INPUT_SELECTOR = exports.EDITING_POPUP_OPTION_NAME = exports.EDITING_FORM_OPTION_NAME = exports.EDITING_EDITROWKEY_OPTION_NAME = exports.EDITING_EDITCOLUMNNAME_OPTION_NAME = exports.DATA_EDIT_DATA_REMOVE_TYPE = exports.DATA_EDIT_DATA_INSERT_TYPE = exports.CELL_MODIFIED_CLASS = void 0;
var EDITOR_CELL_CLASS = "dx-editor-cell";
exports.EDITOR_CELL_CLASS = EDITOR_CELL_CLASS;
var ROW_CLASS = "dx-row";
exports.ROW_CLASS = ROW_CLASS;
var CELL_MODIFIED_CLASS = "dx-cell-modified";
exports.CELL_MODIFIED_CLASS = CELL_MODIFIED_CLASS;
var ROW_SELECTED_CLASS = "dx-selection";
exports.ROW_SELECTED_CLASS = ROW_SELECTED_CLASS;
var EDIT_FORM_CLASS = "edit-form";
exports.EDIT_FORM_CLASS = EDIT_FORM_CLASS;
var DATA_EDIT_DATA_INSERT_TYPE = "insert";
exports.DATA_EDIT_DATA_INSERT_TYPE = DATA_EDIT_DATA_INSERT_TYPE;
var DATA_EDIT_DATA_REMOVE_TYPE = "remove";
exports.DATA_EDIT_DATA_REMOVE_TYPE = DATA_EDIT_DATA_REMOVE_TYPE;
var EDITING_POPUP_OPTION_NAME = "editing.popup";
exports.EDITING_POPUP_OPTION_NAME = EDITING_POPUP_OPTION_NAME;
var EDITING_FORM_OPTION_NAME = "editing.form";
exports.EDITING_FORM_OPTION_NAME = EDITING_FORM_OPTION_NAME;
var EDITING_EDITROWKEY_OPTION_NAME = "editing.editRowKey";
exports.EDITING_EDITROWKEY_OPTION_NAME = EDITING_EDITROWKEY_OPTION_NAME;
var EDITING_EDITCOLUMNNAME_OPTION_NAME = "editing.editColumnName";
exports.EDITING_EDITCOLUMNNAME_OPTION_NAME = EDITING_EDITCOLUMNNAME_OPTION_NAME;
var TARGET_COMPONENT_NAME = "targetComponent";
exports.TARGET_COMPONENT_NAME = TARGET_COMPONENT_NAME;
var EDITORS_INPUT_SELECTOR = "input:not([type='hidden'])";
exports.EDITORS_INPUT_SELECTOR = EDITORS_INPUT_SELECTOR;
var FOCUSABLE_ELEMENT_SELECTOR = "[tabindex], " + EDITORS_INPUT_SELECTOR;
exports.FOCUSABLE_ELEMENT_SELECTOR = FOCUSABLE_ELEMENT_SELECTOR;
var EDIT_MODE_BATCH = "batch";
exports.EDIT_MODE_BATCH = EDIT_MODE_BATCH;
var EDIT_MODE_ROW = "row";
exports.EDIT_MODE_ROW = EDIT_MODE_ROW;
var EDIT_MODE_CELL = "cell";
exports.EDIT_MODE_CELL = EDIT_MODE_CELL;
var EDIT_MODE_FORM = "form";
exports.EDIT_MODE_FORM = EDIT_MODE_FORM;
var EDIT_MODE_POPUP = "popup";
exports.EDIT_MODE_POPUP = EDIT_MODE_POPUP;
var FIRST_NEW_ROW_POSITION = "first";
exports.FIRST_NEW_ROW_POSITION = FIRST_NEW_ROW_POSITION;
var LAST_NEW_ROW_POSITION = "last";
exports.LAST_NEW_ROW_POSITION = LAST_NEW_ROW_POSITION;
var PAGE_BOTTOM_NEW_ROW_POSITION = "pageBottom";
exports.PAGE_BOTTOM_NEW_ROW_POSITION = PAGE_BOTTOM_NEW_ROW_POSITION;
var PAGE_TOP_NEW_ROW_POSITION = "pageTop";
exports.PAGE_TOP_NEW_ROW_POSITION = PAGE_TOP_NEW_ROW_POSITION;
var VIEWPORT_BOTTOM_NEW_ROW_POSITION = "viewportBottom";
exports.VIEWPORT_BOTTOM_NEW_ROW_POSITION = VIEWPORT_BOTTOM_NEW_ROW_POSITION;
var VIEWPORT_TOP_NEW_ROW_POSITION = "viewportTop";
exports.VIEWPORT_TOP_NEW_ROW_POSITION = VIEWPORT_TOP_NEW_ROW_POSITION;
var EDIT_MODES = [EDIT_MODE_BATCH, EDIT_MODE_ROW, EDIT_MODE_CELL, EDIT_MODE_FORM, EDIT_MODE_POPUP];
exports.EDIT_MODES = EDIT_MODES;
var ROW_BASED_MODES = [EDIT_MODE_ROW, EDIT_MODE_FORM, EDIT_MODE_POPUP];
exports.ROW_BASED_MODES = ROW_BASED_MODES;
var MODES_WITH_DELAYED_FOCUS = [EDIT_MODE_ROW, EDIT_MODE_FORM];
exports.MODES_WITH_DELAYED_FOCUS = MODES_WITH_DELAYED_FOCUS;
