// @devexpress/analytics-core/core/utils/parsers.js
/**
* DevExpress Analytics (core\utils\parsers.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var _getNameHelpers_1 = require("../internal/_getNameHelpers");
function floatFromModel(val) {
    return ko.observable(val === undefined || val === null ? null : parseFloat(val));
}
exports.floatFromModel = floatFromModel;
function fromEnum(value) {
    var shotEnumValueKey = _getNameHelpers_1.getShortTypeName(value);
    var valuesArrayItem = this.valuesArray && this.valuesArray.filter(function (item) { return item.value == shotEnumValueKey; })[0];
    return ko.observable((this.values && this.values[shotEnumValueKey] !== undefined || valuesArrayItem) ? shotEnumValueKey : value);
}
exports.fromEnum = fromEnum;
function parseBool(val) {
    return ko.observable(val !== void 0 ? String(val).toLowerCase() === 'true' : val);
}
exports.parseBool = parseBool;
function colorFromString(val) {
    var color = (val || '').split(',');
    var result = ko.observable(val);
    if (color.length === 3) {
        result = ko.observable('rgb(' + color.join(', ') + ')');
    }
    else if (color.length === 4) {
        var alpha = Math.round(parseFloat(color[0]) / 255 * 100) / 100;
        color.shift();
        color.push(alpha.toString());
        result = ko.observable('rgba(' + color.join(', ') + ')');
    }
    return result;
}
exports.colorFromString = colorFromString;
function saveAsInt(val) {
    return Math.round(val).toString();
}
exports.saveAsInt = saveAsInt;
function colorToInt(color) {
    var colorAsString = colorToString(color).split(',');
    return (parseInt(colorAsString[0]) << 24) + (parseInt(colorAsString[1]) << 16) + (parseInt(colorAsString[2]) << 8) + (parseInt(colorAsString[3]));
}
exports.colorToInt = colorToInt;
function intToColor(color, hasAlpha) {
    if (hasAlpha === void 0) { hasAlpha = true; }
    var r = Math.round((color >> 16) & 0xff);
    var g = Math.round((color >> 8) & 0xff);
    var b = Math.round((color) & 0xff);
    if (hasAlpha) {
        var a = Math.round((color >> 24) & 0xff);
        return colorFromString([a, r, g, b].join(', '))();
    }
    else {
        return colorFromString([r, g, b].join(', '))();
    }
}
exports.intToColor = intToColor;
function colorToString(val) {
    var color = (val || '').split(', ');
    var result = val;
    if (color.length === 3) {
        color[0] = color[0].split('(')[1];
        color[2] = color[2].split(')')[0];
        result = color.join(',');
    }
    else if (color.length === 4) {
        var alpha = Math.round(parseFloat(color[3]) * 255);
        color.pop();
        color[0] = color[0].split('(')[1];
        result = alpha.toString() + ',' + color.join(',');
    }
    return result;
}
exports.colorToString = colorToString;
