// @devexpress/analytics-core/core/widgets/_dataMemberEditor.js
/**
* DevExpress Analytics (core\widgets\_dataMemberEditor.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DataMemberTreeListController = (function () {
    function DataMemberTreeListController() {
        this.selectedItem = null;
        this.suppressActions = true;
    }
    DataMemberTreeListController.prototype.dispose = function () {
        this.selectedItem = null;
    };
    DataMemberTreeListController.prototype.itemsFilter = function (item) {
        return item.specifics !== 'parameters' && (item.specifics === 'List' || item.specifics === 'ListSource' || item.isList === true || item.specifics === 'none');
    };
    DataMemberTreeListController.prototype.hasItems = function (item) {
        return item.specifics !== 'none';
    };
    DataMemberTreeListController.prototype.canSelect = function (value) {
        return (this.hasItems(value.data) && !!value.path && (value.data.specifics === 'List' || value.data.specifics === 'ListSource')) || value.data.specifics === 'none';
    };
    DataMemberTreeListController.prototype.select = function (value) {
        if (this.canSelect(value)) {
            this.selectedItem && this.selectedItem.isSelected(false);
            this.selectedItem = value;
            value.isSelected(true);
        }
    };
    return DataMemberTreeListController;
}());
exports.DataMemberTreeListController = DataMemberTreeListController;
