// @devexpress/analytics-core/core/widgets/registerDesignerEditors.js
/**
* DevExpress Analytics (core\widgets\registerDesignerEditors.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var editorsInfo_1 = require("../../property-grid/widgets/editorsInfo");
var guideditor_1 = require("../../property-grid/widgets/guideditor");
var _requiredNullableEditor_1 = require("./_requiredNullableEditor");
var propertygrid_1 = require("../../property-grid/propertygrid");
var fieldListEditor_1 = require("./fieldListEditor");
var dataMemberEditor_1 = require("./dataMemberEditor");
var colorPickerEditor_1 = require("./colorPickerEditor");
function registerDesignerEditors() {
    editorsInfo_1.editorTemplates.registerEditors({
        guid: { header: 'dxrd-guid', editorType: guideditor_1.GuidEditor },
        date: { header: 'dx-date', editorType: _requiredNullableEditor_1.RequiredNullableEditor },
        borders: { header: 'dxrd-borders' },
        textAlignment: { header: 'dxrd-textalignment' },
        objecteditorCustom: { custom: 'dxrd-objectEditorContent', editorType: propertygrid_1.PropertyGridEditor },
        field: { header: 'dxrd-field', editorType: fieldListEditor_1.FieldListEditor },
        dataMember: { header: 'dxrd-field', editorType: dataMemberEditor_1.DataMemberEditor },
        filterEditor: { header: 'dxrd-filterstring' },
        formatEditor: { header: 'dxrd-formatstring' },
        expressionEditor: { header: 'dxrd-expressionstring' },
        customColorEditor: { header: 'dxrd-colorpicker', editorType: colorPickerEditor_1.ColorPickerEditor },
        sbyte: _requiredNullableEditor_1.createNumericEditor('System.SByte', 'integer'),
        decimal: _requiredNullableEditor_1.createNumericEditor('System.Decimal', 'float'),
        int64: _requiredNullableEditor_1.createNumericEditor('System.Int64', 'integer'),
        int32: _requiredNullableEditor_1.createNumericEditor('System.Int32', 'integer'),
        int16: _requiredNullableEditor_1.createNumericEditor('System.Int16', 'integer'),
        single: _requiredNullableEditor_1.createNumericEditor('System.Single', 'float'),
        double: _requiredNullableEditor_1.createNumericEditor('System.Double', 'float'),
        byte: _requiredNullableEditor_1.createNumericEditor('System.Byte', 'integer'),
        uint16: _requiredNullableEditor_1.createNumericEditor('System.UInt16', 'integer'),
        uint32: _requiredNullableEditor_1.createNumericEditor('System.UInt32', 'integer'),
        uint64: _requiredNullableEditor_1.createNumericEditor('System.UInt64', 'integer')
    });
}
exports.registerDesignerEditors = registerDesignerEditors;
