// devextreme/cjs/ui/scheduler/classes.js
/**
 * DevExtreme (cjs/ui/scheduler/classes.js)
 * Version: 21.2.7
 * Build date: Mon Apr 11 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.VIRTUAL_CELL_CLASS = exports.VERTICAL_GROUP_COUNT_CLASSES = exports.TIME_PANEL_CLASS = exports.REDUCED_APPOINTMENT_PARTS_CLASSES = exports.REDUCED_APPOINTMENT_ICON = exports.REDUCED_APPOINTMENT_CLASS = exports.RECURRENCE_APPOINTMENT_CLASS = exports.LAST_GROUP_CELL_CLASS = exports.HEADER_CURRENT_TIME_CELL_CLASS = exports.GROUP_ROW_CLASS = exports.GROUP_HEADER_CONTENT_CLASS = exports.FIXED_CONTAINER_CLASS = exports.FIRST_GROUP_CELL_CLASS = exports.EMPTY_APPOINTMENT_CLASS = exports.DIRECTION_APPOINTMENT_CLASSES = exports.DATE_TABLE_ROW_CLASS = exports.DATE_TABLE_CLASS = exports.APPOINTMENT_ITEM_CLASS = exports.APPOINTMENT_DRAG_SOURCE_CLASS = exports.APPOINTMENT_CONTENT_CLASSES = exports.ALL_DAY_APPOINTMENT_CLASS = exports.AGENDA_LAST_IN_DATE_APPOINTMENT_CLASS = void 0;
var FIXED_CONTAINER_CLASS = "dx-scheduler-fixed-appointments";
exports.FIXED_CONTAINER_CLASS = FIXED_CONTAINER_CLASS;
var REDUCED_APPOINTMENT_CLASS = "dx-scheduler-appointment-reduced";
exports.REDUCED_APPOINTMENT_CLASS = REDUCED_APPOINTMENT_CLASS;
var REDUCED_APPOINTMENT_ICON = "dx-scheduler-appointment-reduced-icon";
exports.REDUCED_APPOINTMENT_ICON = REDUCED_APPOINTMENT_ICON;
var RECURRENCE_APPOINTMENT_CLASS = "dx-scheduler-appointment-recurrence";
exports.RECURRENCE_APPOINTMENT_CLASS = RECURRENCE_APPOINTMENT_CLASS;
var EMPTY_APPOINTMENT_CLASS = "dx-scheduler-appointment-empty";
exports.EMPTY_APPOINTMENT_CLASS = EMPTY_APPOINTMENT_CLASS;
var ALL_DAY_APPOINTMENT_CLASS = "dx-scheduler-all-day-appointment";
exports.ALL_DAY_APPOINTMENT_CLASS = ALL_DAY_APPOINTMENT_CLASS;
var REDUCED_APPOINTMENT_PARTS_CLASSES = {
    head: "dx-scheduler-appointment-head",
    body: "dx-scheduler-appointment-body",
    tail: "dx-scheduler-appointment-tail"
};
exports.REDUCED_APPOINTMENT_PARTS_CLASSES = REDUCED_APPOINTMENT_PARTS_CLASSES;
var DIRECTION_APPOINTMENT_CLASSES = {
    horizontal: "dx-scheduler-appointment-horizontal",
    vertical: "dx-scheduler-appointment-vertical"
};
exports.DIRECTION_APPOINTMENT_CLASSES = DIRECTION_APPOINTMENT_CLASSES;
var APPOINTMENT_DRAG_SOURCE_CLASS = "dx-scheduler-appointment-drag-source";
exports.APPOINTMENT_DRAG_SOURCE_CLASS = APPOINTMENT_DRAG_SOURCE_CLASS;
var APPOINTMENT_ITEM_CLASS = "dx-scheduler-appointment";
exports.APPOINTMENT_ITEM_CLASS = APPOINTMENT_ITEM_CLASS;
var APPOINTMENT_CONTENT_CLASSES = {
    APPOINTMENT_CONTENT_DETAILS: "dx-scheduler-appointment-content-details",
    RECURRING_ICON: "dx-scheduler-appointment-recurrence-icon",
    APPOINTMENT_TITLE: "dx-scheduler-appointment-title",
    APPOINTMENT_DATE: "dx-scheduler-appointment-content-date",
    ALL_DAY_CONTENT: "dx-scheduler-appointment-content-allday",
    ITEM: "dx-scheduler-appointment",
    AGENDA_MARKER: "dx-scheduler-agenda-appointment-marker",
    AGENDA_RESOURCE_LIST: "dx-scheduler-appointment-resource-list",
    AGENDA_RESOURCE_LIST_ITEM: "dx-scheduler-appointment-resource-item",
    AGENDA_RESOURCE_LIST_ITEM_VALUE: "dx-scheduler-appointment-resource-item-value"
};
exports.APPOINTMENT_CONTENT_CLASSES = APPOINTMENT_CONTENT_CLASSES;
var AGENDA_LAST_IN_DATE_APPOINTMENT_CLASS = "dx-scheduler-last-in-date-agenda-appointment";
exports.AGENDA_LAST_IN_DATE_APPOINTMENT_CLASS = AGENDA_LAST_IN_DATE_APPOINTMENT_CLASS;
var HEADER_CURRENT_TIME_CELL_CLASS = "dx-scheduler-header-panel-current-time-cell";
exports.HEADER_CURRENT_TIME_CELL_CLASS = HEADER_CURRENT_TIME_CELL_CLASS;
var VIRTUAL_CELL_CLASS = "dx-scheduler-virtual-cell";
exports.VIRTUAL_CELL_CLASS = VIRTUAL_CELL_CLASS;
var TIME_PANEL_CLASS = "dx-scheduler-time-panel";
exports.TIME_PANEL_CLASS = TIME_PANEL_CLASS;
var DATE_TABLE_CLASS = "dx-scheduler-date-table";
exports.DATE_TABLE_CLASS = DATE_TABLE_CLASS;
var DATE_TABLE_ROW_CLASS = "dx-scheduler-date-table-row";
exports.DATE_TABLE_ROW_CLASS = DATE_TABLE_ROW_CLASS;
var GROUP_ROW_CLASS = "dx-scheduler-group-row";
exports.GROUP_ROW_CLASS = GROUP_ROW_CLASS;
var GROUP_HEADER_CONTENT_CLASS = "dx-scheduler-group-header-content";
exports.GROUP_HEADER_CONTENT_CLASS = GROUP_HEADER_CONTENT_CLASS;
var LAST_GROUP_CELL_CLASS = "dx-scheduler-last-group-cell";
exports.LAST_GROUP_CELL_CLASS = LAST_GROUP_CELL_CLASS;
var FIRST_GROUP_CELL_CLASS = "dx-scheduler-first-group-cell";
exports.FIRST_GROUP_CELL_CLASS = FIRST_GROUP_CELL_CLASS;
var VERTICAL_GROUP_COUNT_CLASSES = ["dx-scheduler-group-column-count-one", "dx-scheduler-group-column-count-two", "dx-scheduler-group-column-count-three"];
exports.VERTICAL_GROUP_COUNT_CLASSES = VERTICAL_GROUP_COUNT_CLASSES;
