// devextreme/cjs/ui/scheduler/constants.js
/**
 * DevExtreme (cjs/ui/scheduler/constants.js)
 * Version: 21.2.7
 * Build date: Mon Apr 11 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.VIEWS = exports.LIST_ITEM_DATA_KEY = exports.LIST_ITEM_CLASS = exports.HORIZONTAL_GROUP_ORIENTATION = exports.APPOINTMENT_SETTINGS_KEY = void 0;
var LIST_ITEM_DATA_KEY = "dxListItemData";
exports.LIST_ITEM_DATA_KEY = LIST_ITEM_DATA_KEY;
var LIST_ITEM_CLASS = "dx-list-item";
exports.LIST_ITEM_CLASS = LIST_ITEM_CLASS;
var APPOINTMENT_SETTINGS_KEY = "dxAppointmentSettings";
exports.APPOINTMENT_SETTINGS_KEY = APPOINTMENT_SETTINGS_KEY;
var HORIZONTAL_GROUP_ORIENTATION = "horizontal";
exports.HORIZONTAL_GROUP_ORIENTATION = HORIZONTAL_GROUP_ORIENTATION;
var VIEWS = {
    DAY: "day",
    WEEK: "week",
    WORK_WEEK: "workWeek",
    MONTH: "month",
    TIMELINE_DAY: "timelineDay",
    TIMELINE_WEEK: "timelineWeek",
    TIMELINE_WORK_WEEK: "timelineWorkWeek",
    TIMELINE_MONTH: "timelineMonth",
    AGENDA: "agenda"
};
exports.VIEWS = VIEWS;
