// @devexpress/analytics-core/query-builder/wizard/fullscreen/pages/specifyJsonDataSourceSettingsPage.js
/**
* DevExpress Analytics (query-builder\wizard\fullscreen\pages\specifyJsonDataSourceSettingsPage.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var fullscreenWizardPage_1 = require("./fullscreenWizardPage");
var specifyJsonConnectionPage_1 = require("../../pages/jsonDataSourceWizard/specifyJsonConnectionPage");
var pageId_1 = require("../../pageId");
var chooseJsonSourcePage_1 = require("../../pages/jsonDataSourceWizard/chooseJsonSourcePage");
var localization_utils_1 = require("../../../../property-grid/localization/localization_utils");
var chooseJsonSchemaPage_1 = require("../../pages/jsonDataSourceWizard/chooseJsonSchemaPage");
var SpecifyJsonDataSourceSettingsPage = (function (_super) {
    __extends(SpecifyJsonDataSourceSettingsPage, _super);
    function SpecifyJsonDataSourceSettingsPage(_dataSourceWizardOptions) {
        var _this = _super.call(this) || this;
        _this._dataSourceWizardOptions = _dataSourceWizardOptions;
        return _this;
    }
    SpecifyJsonDataSourceSettingsPage.prototype.registerSections = function () {
        if (this._dataSourceWizardOptions.connectionStrings.json().length > 0) {
            specifyJsonConnectionPage_1._registerSpecifyJsonConnectionPage(this._factory, this._dataSourceWizardOptions.connectionStrings.json, this._dataSourceWizardOptions.allowCreateNewJsonConnection, this._dataSourceWizardOptions.callbacks.getItemsProviderCallback);
            this._setSectionPosition(pageId_1.FullscreenDataSourceWizardSectionId.SpecifyJsonConnectionPage, this._dataSourceWizardOptions.rtl ? fullscreenWizardPage_1.WizardSectionPosition.Right : fullscreenWizardPage_1.WizardSectionPosition.Left);
        }
        if (this._dataSourceWizardOptions.allowCreateNewJsonConnection) {
            if (this._dataSourceWizardOptions.connectionStrings.json().length === 0) {
                chooseJsonSourcePage_1._registerChooseJsonSourcePage(this._factory, this._dataSourceWizardOptions.requestWrapper, this._dataSourceWizardOptions.callbacks.getItemsProviderCallback);
                this._setSectionPosition(pageId_1.FullscreenDataSourceWizardSectionId.ChooseJsonSourcePage, this._dataSourceWizardOptions.rtl ? fullscreenWizardPage_1.WizardSectionPosition.Right : fullscreenWizardPage_1.WizardSectionPosition.Left);
                var meta = this._factory.getMetadata(pageId_1.FullscreenDataSourceWizardSectionId.ChooseJsonSourcePage);
                meta['disabledText'] = localization_utils_1.getLocalization("To create a data connection, select \"No, I'd like to create a new data connection\".", 'AnalyticsCoreStringId.JsonDSWizard_CreateNewConnectionPage_Placeholder');
            }
        }
        chooseJsonSchemaPage_1._registerChooseJsonSchemaPage(this._factory, this._dataSourceWizardOptions.requestWrapper, this._dataSourceWizardOptions.callbacks);
        this._setSectionPosition(pageId_1.FullscreenDataSourceWizardSectionId.ChooseJsonSchemaPage, this._dataSourceWizardOptions.rtl ? fullscreenWizardPage_1.WizardSectionPosition.Left : fullscreenWizardPage_1.WizardSectionPosition.Right);
        var meta = this._factory.getMetadata(pageId_1.FullscreenDataSourceWizardSectionId.ChooseJsonSchemaPage);
        meta['disabledText'] = localization_utils_1.getLocalization('To select data fields, choose or create a data connection.', 'AnalyticsCoreStringId.JsonDSWizard_ChooseJsonSchemaPage_Placeholder');
    };
    SpecifyJsonDataSourceSettingsPage.prototype.getNextSectionId = function (sectionId) {
        if (!sectionId && this._dataSourceWizardOptions.connectionStrings.json().length > 0) {
            return pageId_1.FullscreenDataSourceWizardSectionId.SpecifyJsonConnectionPage;
        }
        else if (!sectionId) {
            return pageId_1.FullscreenDataSourceWizardSectionId.ChooseJsonSourcePage;
        }
        else if (this._dataSourceWizardOptions.allowCreateNewJsonConnection) {
            if (sectionId === pageId_1.FullscreenDataSourceWizardSectionId.SpecifyJsonConnectionPage)
                return pageId_1.FullscreenDataSourceWizardSectionId.ChooseJsonSchemaPage;
            else if (sectionId === pageId_1.FullscreenDataSourceWizardSectionId.ChooseJsonSourcePage)
                return pageId_1.FullscreenDataSourceWizardSectionId.ChooseJsonSchemaPage;
        }
        else {
            if (sectionId === pageId_1.FullscreenDataSourceWizardSectionId.SpecifyJsonConnectionPage)
                return pageId_1.FullscreenDataSourceWizardSectionId.ChooseJsonSchemaPage;
        }
    };
    SpecifyJsonDataSourceSettingsPage.prototype.canNext = function () {
        var section = this._sectionsProcessor.getPageById(pageId_1.FullscreenDataSourceWizardSectionId.ChooseJsonSchemaPage);
        return section && section.page() && section.metadata.canNext(section.page().page);
    };
    return SpecifyJsonDataSourceSettingsPage;
}(fullscreenWizardPage_1.FullscreenWizardPage));
exports.SpecifyJsonDataSourceSettingsPage = SpecifyJsonDataSourceSettingsPage;
function _registerSpecifyJsonDataSourceSettingsPage(factory, dataSourceWizardOptions) {
    factory.registerMetadata(pageId_1.FullscreenDataSourceWizardPageId.SpecifyJsonDataSourceSettingsPage, {
        setState: function (data, state) {
            state.jsonDataSourceWizard.connectionName = data.jsonDataSourceWizard.connectionName;
            state.jsonDataSourceWizard.newConnectionName = data.jsonDataSourceWizard.newConnectionName;
            state.jsonDataSourceWizard.jsonSource = data.jsonDataSourceWizard.jsonSource;
            state.jsonDataSourceWizard.dataSourceName = data.jsonDataSourceWizard.dataSourceName;
            state.jsonDataSourceWizard.jsonScheme = data.jsonDataSourceWizard.jsonScheme;
            state.jsonDataSourceWizard.rootElement = data.jsonDataSourceWizard.rootElement;
        },
        getState: function (state) { return state; },
        resetState: function (state, defaultState) {
            state.jsonDataSourceWizard.connectionName = defaultState.jsonDataSourceWizard.connectionName;
            state.jsonDataSourceWizard.jsonSource = defaultState.jsonDataSourceWizard.jsonSource;
            state.jsonDataSourceWizard.dataSourceName = defaultState.jsonDataSourceWizard.dataSourceName;
            state.jsonDataSourceWizard.newConnectionName = defaultState.jsonDataSourceWizard.newConnectionName;
            state.jsonDataSourceWizard.jsonScheme = defaultState.jsonDataSourceWizard.jsonScheme;
            state.jsonDataSourceWizard.rootElement = defaultState.jsonDataSourceWizard.rootElement;
        },
        create: function () {
            return new SpecifyJsonDataSourceSettingsPage(dataSourceWizardOptions);
        },
        navigationPanelText: localization_utils_1.getLocalization('Specify Data Source Settings', 'AnalyticsCoreStringId.Wizard_SpecifyDataSourceSettingsPage'),
        template: 'dx-wizard-fullscreen-page'
    });
}
exports._registerSpecifyJsonDataSourceSettingsPage = _registerSpecifyJsonDataSourceSettingsPage;
