// @devexpress/analytics-core/query-builder/wizard/fullscreen/pages/specifySqlDataSourceSettingsPage.js
/**
* DevExpress Analytics (query-builder\wizard\fullscreen\pages\specifySqlDataSourceSettingsPage.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var fullscreenWizardPage_1 = require("./fullscreenWizardPage");
var pageId_1 = require("../../pageId");
var dataSourceWizardState_1 = require("../../dataSourceWizardState");
var chooseSqlConnectionPage_1 = require("../../pages/chooseSqlConnectionPage");
var multiQueryConfigurePage_1 = require("../../pages/multiQueryWizard/multiQueryConfigurePage");
var configureMasterDetailRelationshipsPage_1 = require("../../pages/multiQueryWizard/configureMasterDetailRelationshipsPage");
var _requestwrapper_1 = require("../../../utils/_requestwrapper");
var multiQueryConfigureParametersPage_1 = require("../../pages/multiQueryWizard/multiQueryConfigureParametersPage");
var localization_utils_1 = require("../../../../property-grid/localization/localization_utils");
var _utils_1 = require("../../../../serializer/_utils");
var SpecifySqlDataSourceSettingsPage = (function (_super) {
    __extends(SpecifySqlDataSourceSettingsPage, _super);
    function SpecifySqlDataSourceSettingsPage(_dataSourceWizardOptions) {
        var _this = _super.call(this) || this;
        _this._dataSourceWizardOptions = _dataSourceWizardOptions;
        return _this;
    }
    SpecifySqlDataSourceSettingsPage.prototype.getNextSectionId = function (sectionId) {
        if (!sectionId)
            return pageId_1.FullscreenDataSourceWizardSectionId.ChooseSqlConnectionPage;
        else if (sectionId === pageId_1.FullscreenDataSourceWizardSectionId.ChooseSqlConnectionPage && this._stateManager.getCurrentState().sqlDataSourceWizard.name) {
            return pageId_1.FullscreenDataSourceWizardSectionId.ConfigureQueryPage;
        }
        else if (sectionId === pageId_1.FullscreenDataSourceWizardSectionId.ConfigureQueryPage) {
            var sections = [];
            var sqlDataSourceWizardState = this._stateManager.getCurrentState().sqlDataSourceWizard;
            if ((sqlDataSourceWizardState.customQueries || []).length > 0) {
                sections.push(pageId_1.FullscreenDataSourceWizardSectionId.ConfigureQueryParametersPage);
            }
            var sqlDataSourceWrapped = dataSourceWizardState_1._restoreSqlDataSourceFromState(sqlDataSourceWizardState);
            if (sqlDataSourceWrapped.sqlDataSource.queries().length > 1) {
                sections.push(pageId_1.FullscreenDataSourceWizardSectionId.ConfigureMasterDetailRelationshipsPage);
            }
            return sections;
        }
    };
    SpecifySqlDataSourceSettingsPage.prototype.registerSections = function () {
        chooseSqlConnectionPage_1._registerChooseSqlConnectionPage(this._factory, this._dataSourceWizardOptions.connectionStrings.sql);
        multiQueryConfigurePage_1._registerMultiQueryConfigurePage(this._factory, this._dataSourceWizardOptions);
        configureMasterDetailRelationshipsPage_1._registerConfigureMasterDetailRelationshipsPage(this._factory, _requestwrapper_1.wrapRebuildResultSchema(this._dataSourceWizardOptions.callbacks.sqlDataSourceResultSchema));
        multiQueryConfigureParametersPage_1._registerMultiQueryConfigureParametersPage(this._factory, this._dataSourceWizardOptions.requestWrapper);
        var meta = this._factory.getMetadata(pageId_1.FullscreenDataSourceWizardSectionId.ChooseSqlConnectionPage);
        meta.description = localization_utils_1.getLocalization('Choose a data connection.', 'AnalyticsCoreStringId.SqlDSWizard_PageChooseConnection');
        var meta = this._factory.getMetadata(pageId_1.FullscreenDataSourceWizardSectionId.ConfigureQueryPage);
        meta['recreate'] = true;
        meta.description = localization_utils_1.getLocalization('Choose predefined queries and/or create custom queries.', 'AnalyticsCoreStringId.Wizard_Queries_Description');
        meta['required'] = true;
        var meta = this._factory.getMetadata(pageId_1.FullscreenDataSourceWizardSectionId.ConfigureMasterDetailRelationshipsPage);
        meta.description = localization_utils_1.getLocalization('Configure master-detail relationships.', 'AnalyticsCoreStringId.SqlDSWizard_PageConfigureMasterDetailRelations');
        meta['disabledText'] = localization_utils_1.getLocalization('To create a master-detail relationship, select two or more queries.', 'AnalyticsCoreStringId.Wizard_MasterDetailRelationship_Placeholder');
        var meta = this._factory.getMetadata(pageId_1.FullscreenDataSourceWizardSectionId.ConfigureQueryParametersPage);
        meta.description = localization_utils_1.getLocalization('Configure query parameters.', 'AnalyticsCoreStringId.SqlDSWizard_PageConfigureParameters');
        meta['disabledText'] = localization_utils_1.getLocalization('To specify query parameters, select a parameterized stored procedure or create a custom query.', 'AnalyticsCoreStringId.Wizard_ConfigureQueryParameters_Placeholder');
        this._setSectionPosition(pageId_1.FullscreenDataSourceWizardSectionId.ChooseSqlConnectionPage, this._dataSourceWizardOptions.rtl ? fullscreenWizardPage_1.WizardSectionPosition.TopRight : fullscreenWizardPage_1.WizardSectionPosition.TopLeft);
        this._setSectionPosition(pageId_1.FullscreenDataSourceWizardSectionId.ConfigureQueryPage, this._dataSourceWizardOptions.rtl ? fullscreenWizardPage_1.WizardSectionPosition.TopLeft : fullscreenWizardPage_1.WizardSectionPosition.TopRight);
        this._setSectionPosition(pageId_1.FullscreenDataSourceWizardSectionId.ConfigureQueryParametersPage, this._dataSourceWizardOptions.rtl ? fullscreenWizardPage_1.WizardSectionPosition.BottomLeft : fullscreenWizardPage_1.WizardSectionPosition.BottomRight);
        this._setSectionPosition(pageId_1.FullscreenDataSourceWizardSectionId.ConfigureMasterDetailRelationshipsPage, this._dataSourceWizardOptions.rtl ? fullscreenWizardPage_1.WizardSectionPosition.BottomRight : fullscreenWizardPage_1.WizardSectionPosition.BottomLeft);
    };
    return SpecifySqlDataSourceSettingsPage;
}(fullscreenWizardPage_1.FullscreenWizardPage));
exports.SpecifySqlDataSourceSettingsPage = SpecifySqlDataSourceSettingsPage;
function _registerSpecifySqlDataSourceSettingsPage(factory, dataSourceWizardOptions) {
    factory.registerMetadata(pageId_1.FullscreenDataSourceWizardPageId.SpecifySqlDataSourceSettingsPage, {
        setState: function (data, state) {
            state.sqlDataSourceWizard.customQueries = data.sqlDataSourceWizard.customQueries;
            state.sqlDataSourceWizard.name = data.sqlDataSourceWizard.name;
            state.sqlDataSourceWizard.queryName = data.sqlDataSourceWizard.name;
            state.sqlDataSourceWizard.sqlDataSourceJSON = data.sqlDataSourceWizard.sqlDataSourceJSON;
            state.sqlDataSourceWizard.relations = data.sqlDataSourceWizard.relations;
        },
        getState: function (state) { return state; },
        resetState: function (state, defaulState) {
            state.sqlDataSourceWizard = _utils_1.extend(true, {}, defaulState);
        },
        create: function () {
            return new SpecifySqlDataSourceSettingsPage(dataSourceWizardOptions);
        },
        navigationPanelText: localization_utils_1.getLocalization('Specify Data Source Settings', 'AnalyticsCoreStringId.Wizard_SpecifyDataSourceSettingsPage'),
        template: 'dx-wizard-fullscreen-page'
    });
}
exports._registerSpecifySqlDataSourceSettingsPage = _registerSpecifySqlDataSourceSettingsPage;
