// @devexpress/analytics-core/query-builder/wizard/fullscreen/pages/specifyObjectDataSourceSettingsPage.js
/**
* DevExpress Analytics (query-builder\wizard\fullscreen\pages\specifyObjectDataSourceSettingsPage.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var $ = require("jquery");
var _utils_1 = require("../../internal/_utils");
var wizardPage_1 = require("../../pages/wizardPage");
var objectSchema_1 = require("../../../dataSource/object/objectSchema");
var objectDataSource_1 = require("../../../dataSource/object/objectDataSource");
var _objectSchemaProvider_1 = require("../../internal/objectDataSource/_objectSchemaProvider");
var _chooseObjectTypes_1 = require("../../internal/objectDataSource/_chooseObjectTypes");
var _chooseObjectDataMembers_1 = require("../../internal/objectDataSource/_chooseObjectDataMembers");
var _chooseObjectParameters_1 = require("../../internal/objectDataSource/_chooseObjectParameters");
var localization_utils_1 = require("../../../../property-grid/localization/localization_utils");
var fullscreenWizardPage_1 = require("./fullscreenWizardPage");
var pageId_1 = require("../../pageId");
var SpecifyObjectDataSourceSettingsPage = (function (_super) {
    __extends(SpecifyObjectDataSourceSettingsPage, _super);
    function SpecifyObjectDataSourceSettingsPage(_dataSourceWizardOptions) {
        var _this = _super.call(this) || this;
        _this._dataSourceWizardOptions = _dataSourceWizardOptions;
        _this._types = ko.observableArray([]);
        _this._objectDataSource = new objectDataSource_1.ObjectDataSource();
        var getItemsProviderCallback = _this._dataSourceWizardOptions.callbacks.getItemsProviderCallback;
        _this._disposables.push(_this._provider = new _objectSchemaProvider_1.ObjectSchemaProvider(_this._dataSourceWizardOptions.requestWrapper), _this._chooseObjectType = new _chooseObjectTypes_1.ChooseObjectTypes(_this._types, _this._provider), _this._chooseObjectDataMember = new _chooseObjectDataMembers_1.ChooseObjectDataMembers(_this._chooseObjectType.selectedType, _this._chooseObjectType.selectedCtor), _this._chooseObjectParameters = new _chooseObjectParameters_1.ChooseObjectParameters(_this._chooseObjectType.selectedCtor, _this._chooseObjectDataMember.selectedDataMembers, getItemsProviderCallback && getItemsProviderCallback()));
        _this._initSections();
        return _this;
    }
    SpecifyObjectDataSourceSettingsPage.prototype._initSections = function () {
        var _this = this;
        this._sections = [
            {
                data: this._chooseObjectType,
                disabled: function () { return false; },
                description: localization_utils_1.getLocalization('Choose the type and its constructor.', 'AnalyticsCoreStringId.ObjectDSWizard_ChooseType_Description'),
                position: _utils_1.getSectionStyle(this._dataSourceWizardOptions.rtl ? fullscreenWizardPage_1.WizardSectionPosition.Right : fullscreenWizardPage_1.WizardSectionPosition.Left),
                template: 'dx-objectdatasource-types-section'
            },
            {
                data: this._chooseObjectDataMember,
                disabled: function () { return !_this._chooseObjectDataMember.dataMembers().length; },
                description: localization_utils_1.getLocalization('Choose the entire object or a data member to bind.', 'AnalyticsCoreStringId.ObjectDSWizard_ChooseDataMember_Description'),
                position: _utils_1.getSectionStyle(this._dataSourceWizardOptions.rtl ? fullscreenWizardPage_1.WizardSectionPosition.TopLeft : fullscreenWizardPage_1.WizardSectionPosition.TopRight),
                disabledText: localization_utils_1.getLocalization('To select a data member, choose a type that contains at least one data member.', 'AnalyticsCoreStringId.ObjectDSWizard_ChooseDataMember_Disabled_Description'),
                template: 'dx-objectdatasource-datamembers-section'
            },
            {
                data: this._chooseObjectParameters,
                disabled: function () { return !_this._chooseObjectParameters.hasParameters(); },
                position: _utils_1.getSectionStyle(this._dataSourceWizardOptions.rtl ? fullscreenWizardPage_1.WizardSectionPosition.BottomLeft : fullscreenWizardPage_1.WizardSectionPosition.BottomRight),
                description: localization_utils_1.getLocalization('Configure constructor parameters and/or method parameters.', 'AnalyticsCoreStringId.ObjectDSWizard_ConfigureParameters_Description'),
                disabledText: localization_utils_1.getLocalization('To specify parameters, select a parameterized constructor or method.', 'AnalyticsCoreStringId.ObjectDSWizard_ConfigureParameters_Disabled_Description'),
                template: 'dx-objectdatasource-parameters-section'
            }
        ];
    };
    SpecifyObjectDataSourceSettingsPage.prototype.showDescription = function (index, text) {
        return [index + 1, text].join('. ');
    };
    SpecifyObjectDataSourceSettingsPage.prototype.canNext = function () {
        return false;
    };
    SpecifyObjectDataSourceSettingsPage.prototype.canFinish = function () {
        return !!this._chooseObjectDataMember.selectedDataMembers().length;
    };
    SpecifyObjectDataSourceSettingsPage.prototype.commit = function () {
        var selectedDataMember = this._chooseObjectDataMember.selectedDataMembers()[0];
        return $.Deferred().resolve({
            selectedType: this._chooseObjectType.selectedType().name,
            ctor: this._chooseObjectType.selectedCtor(),
            dataSourceName: this._objectDataSource.name(),
            context: this._context,
            dataMember: !selectedDataMember || selectedDataMember.isEntireObject() ? undefined : selectedDataMember,
        }).promise();
    };
    SpecifyObjectDataSourceSettingsPage.prototype.initialize = function (state) {
        var _this = this;
        this._context = state.objectDataSourceWizard.context;
        this._objectDataSource.setState(state.objectDataSourceWizard);
        this._chooseObjectType.selectedCtor(this._objectDataSource.ctor);
        this._chooseObjectDataMember.selectedDataMembers([this._objectDataSource.dataMember]);
        this._provider.getObjectTypeDescriptions(state.objectDataSourceWizard.context).done(function (schema) {
            schema.types.forEach(function (type) { return type.members.splice(0, 0, objectSchema_1.ObjectDataMember.empty()); });
            _this._types(schema.types || []);
            if (schema.types.length > 0) {
                _this._chooseObjectType.selectedPath(schema.types[0].name);
            }
        });
        return $.Deferred().resolve().promise();
    };
    return SpecifyObjectDataSourceSettingsPage;
}(wizardPage_1.WizardPageBase));
exports.SpecifyObjectDataSourceSettingsPage = SpecifyObjectDataSourceSettingsPage;
function _registerSpecifyObjectDataSourceSettingsPage(factory, dataSourceWizardOptions) {
    factory.registerMetadata(pageId_1.FullscreenDataSourceWizardPageId.SpecifyObjectDataSourceSettingsPage, {
        setState: function (data, state) {
            state.objectDataSourceWizard.ctor = data.ctor;
            state.objectDataSourceWizard.dataMember = data.dataMember;
            state.objectDataSourceWizard.selectedType = data.selectedType;
            state.objectDataSourceWizard.dataSourceName = data.dataSourceName;
        },
        getState: function (state) {
            return state;
        },
        resetState: function (state, defaultState) {
            state.objectDataSourceWizard.ctor = defaultState.objectDataSourceWizard.ctor;
            state.objectDataSourceWizard.dataMember = defaultState.objectDataSourceWizard.dataMember;
            state.objectDataSourceWizard.selectedType = defaultState.objectDataSourceWizard.selectedType;
            state.objectDataSourceWizard.dataSourceName = defaultState.objectDataSourceWizard.dataSourceName;
        },
        create: function () {
            return new SpecifyObjectDataSourceSettingsPage(dataSourceWizardOptions);
        },
        template: 'dxrd-page-objectsource',
        navigationPanelText: localization_utils_1.getLocalization('Specify Data Source Settings', 'AnalyticsCoreStringId.Wizard_SpecifyDataSourceSettingsPage')
    });
}
exports._registerSpecifyObjectDataSourceSettingsPage = _registerSpecifyObjectDataSourceSettingsPage;
