// @devexpress/analytics-core/query-builder/wizard/pages/objectDataSourceWizard/chooseObjectDataSourceDataMembersPage.js
/**
* DevExpress Analytics (query-builder\wizard\pages\objectDataSourceWizard\chooseObjectDataSourceDataMembersPage.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var $ = require("jquery");
var wizardPage_1 = require("../wizardPage");
var requestwrapper_1 = require("../../../utils/requestwrapper");
var _chooseObjectDataMembers_1 = require("../../internal/objectDataSource/_chooseObjectDataMembers");
var objectDataSource_1 = require("../../../dataSource/object/objectDataSource");
var pageId_1 = require("../../pageId");
var localization_utils_1 = require("../../../../property-grid/localization/localization_utils");
var ChooseObjectDataSourceDataMembersPage = (function (_super) {
    __extends(ChooseObjectDataSourceDataMembersPage, _super);
    function ChooseObjectDataSourceDataMembersPage(_requestWrapper) {
        if (_requestWrapper === void 0) { _requestWrapper = new requestwrapper_1.RequestWrapper(); }
        var _this = _super.call(this) || this;
        _this._requestWrapper = _requestWrapper;
        _this._objectDataSource = new objectDataSource_1.ObjectDataSource();
        _this._type = ko.observable();
        _this._ctor = ko.observable();
        _this._disposables.push(_this._chooseObjectDataMember = new _chooseObjectDataMembers_1.ChooseObjectDataMembers(_this._type, _this._ctor));
        return _this;
    }
    ChooseObjectDataSourceDataMembersPage.prototype.initialize = function (state) {
        this._objectDataSource.setState(state);
        this._type(state.selectedObjectType);
        this._ctor(state.ctor);
        return $.Deferred().resolve().promise();
    };
    ChooseObjectDataSourceDataMembersPage.prototype.canNext = function () {
        return !!this._needParametersPage;
    };
    ChooseObjectDataSourceDataMembersPage.prototype.canFinish = function () {
        return !this._needParametersPage;
    };
    ChooseObjectDataSourceDataMembersPage.prototype.commit = function () {
        return $.Deferred().resolve({
            selectedType: this._type().name,
            ctor: this._ctor(),
            dataMember: !this._selectedDataMember || this._selectedDataMember.isEntireObject() ? undefined : this._selectedDataMember,
        }).promise();
    };
    Object.defineProperty(ChooseObjectDataSourceDataMembersPage.prototype, "_selectedDataMember", {
        get: function () {
            var selectedDataMembers = this._chooseObjectDataMember.selectedDataMembers();
            return !!selectedDataMembers ? selectedDataMembers[0] : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ChooseObjectDataSourceDataMembersPage.prototype, "_needParametersPage", {
        get: function () {
            return (!!this._selectedDataMember && this._selectedDataMember.parameters.length > 0) ||
                (!!this._ctor() && this._ctor().parameters.length > 0);
        },
        enumerable: true,
        configurable: true
    });
    return ChooseObjectDataSourceDataMembersPage;
}(wizardPage_1.WizardPageBase));
exports.ChooseObjectDataSourceDataMembersPage = ChooseObjectDataSourceDataMembersPage;
function _registerChooseObjectDataSourceDataMembersPage(factory, dataSourceWizardOptions) {
    factory.registerMetadata(pageId_1.ObjectDataSourceWizardPageId.ChooseDataMembersPage, {
        setState: function (data, state) {
            state.selectedType = data.selectedType;
            state.ctor = data.ctor;
            state.dataMember = data.dataMember;
        },
        getState: function (state) {
            return state.objectDataSourceWizard;
        },
        resetState: function (state, defaultState) {
            state.dataMember = defaultState.dataMember;
        },
        create: function () {
            return new ChooseObjectDataSourceDataMembersPage(dataSourceWizardOptions.requestWrapper);
        },
        description: localization_utils_1.getLocalization('Choose the entire object or a data member to bind.', 'AnalyticsCoreStringId.ObjectDSWizard_ChooseDataMember_Description'),
        template: 'dxrd-page-objectdatasource-datamembers'
    });
}
exports._registerChooseObjectDataSourceDataMembersPage = _registerChooseObjectDataSourceDataMembersPage;
