// @devexpress/analytics-core/widgets/ace/_ace-mode-doc-comment.js
/**
* DevExpress Analytics (widgets\ace\_ace-mode-doc-comment.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ace_1 = require("ace-builds/src-noconflict/ace");
if (!!ace_1.default) {
    ace_1.default.define('ace/mode/doc_comment_highlight_rules', ['require', 'exports', 'module', 'ace/lib/oop', 'ace/mode/text_highlight_rules'], function (require, exports, module) {
        'use strict';
        var oop = ace_1.default.require('ace/lib/oop');
        var TextHighlightRules = ace_1.default.require('ace/mode/text_highlight_rules').TextHighlightRules;
        var DocCommentHighlightRules = function () {
            this.$rules = {
                'start': [{
                        token: 'comment.doc.tag',
                        regex: '@[\\w\\d_]+'
                    },
                    DocCommentHighlightRules['getTagRule'](),
                    {
                        defaultToken: 'comment.doc',
                        caseInsensitive: true
                    }]
            };
        };
        oop.inherits(DocCommentHighlightRules, TextHighlightRules);
        DocCommentHighlightRules['getTagRule'] = function (start) {
            return {
                token: 'comment.doc.tag.storage.type',
                regex: '\\b(?:TODO|FIXME|XXX|HACK)\\b'
            };
        };
        DocCommentHighlightRules['getStartRule'] = function (start) {
            return {
                token: 'comment.doc',
                regex: '\\/\\*(?=\\*)',
                next: start
            };
        };
        DocCommentHighlightRules['getEndRule'] = function (start) {
            return {
                token: 'comment.doc',
                regex: '\\*\\/',
                next: start
            };
        };
        exports.DocCommentHighlightRules = DocCommentHighlightRules;
    });
}
