// @devexpress/analytics-core/core/internal/_getNameHelpers.js
/**
* DevExpress Analytics (core\internal\_getNameHelpers.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var localization_utils_1 = require("../../property-grid/localization/localization_utils");
function getTypeNameFromFullName(controlType) {
    return controlType.split(',')[0].trim();
}
exports.getTypeNameFromFullName = getTypeNameFromFullName;
function getShortTypeName(controlType) {
    var fullTypeName = getTypeNameFromFullName(controlType), typeNameParts = fullTypeName.split('.');
    return typeNameParts[typeNameParts.length - 1];
}
exports.getShortTypeName = getShortTypeName;
function getControlTypeName(value) {
    var controlType = value.displayType && value.displayType() || value.controlType;
    if (value.getControlFactory) {
        var info = value.getControlFactory().getControlInfo(controlType);
        if (info.displayName)
            controlType = info.displayName;
    }
    return controlType && localization_utils_1.getLocalization(getShortTypeName(controlType));
}
exports.getControlTypeName = getControlTypeName;
function getControlFullName(value) {
    if (!value)
        return '';
    var displayName = ko.unwrap(value.name) || ko.unwrap(value.displayName), controlType = getControlTypeName(value);
    return displayName + (controlType ? (' (' + controlType + ')') : '');
}
exports.getControlFullName = getControlFullName;
function getImageClassName(controlType, isTemplate) {
    if (isTemplate === void 0) { isTemplate = false; }
    var controlType = getTypeNameFromFullName(controlType || '').split('.').join('_'), name;
    if (controlType.indexOf('XR') === 0) {
        name = controlType.slice(2).toLowerCase();
    }
    else if (controlType === 'DevExpress_XtraReports_UI_XtraReport') {
        name = 'master_report';
    }
    return (isTemplate ? 'dxrd-svg-toolbox-' : 'dxrd-image-') + (name ? name : controlType.toLowerCase());
}
exports.getImageClassName = getImageClassName;
function getUniqueNameForNamedObjectsArray(objects, prefix, names) {
    if (prefix.indexOf('XR') === 0) {
        prefix = prefix[2].toLowerCase() + prefix.slice(3);
    }
    else {
        var indexBand = prefix.indexOf('Band');
        if (indexBand !== -1 && prefix !== 'SubBand') {
            prefix = prefix.slice(0, indexBand) + prefix.slice(indexBand + 4);
        }
    }
    return getUniqueName(names || objects.map(function (item) { return ko.unwrap(item.name); }), prefix);
}
exports.getUniqueNameForNamedObjectsArray = getUniqueNameForNamedObjectsArray;
function getUniqueName(names, prefix) {
    var i = 1, result = prefix + i;
    while (names.filter(function (item) { return item === result; }).length > 0) {
        i++;
        result = prefix + i;
    }
    return result;
}
exports.getUniqueName = getUniqueName;
