// @devexpress/analytics-core/core/selection/_combinedObj.js
/**
* DevExpress Analytics (core\selection\_combinedObj.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var $ = require("jquery");
var _utils_1 = require("../../serializer/_utils");
var CombinedObject = (function () {
    function CombinedObject() {
    }
    CombinedObject.getInfo = function (controls) {
        var infos = controls.map(function (item) { return item['getInfo'](); });
        return function () {
            var info = [];
            for (var i = 0; i < infos[0].length; i++) {
                if (infos.filter(function (info) { return info.filter(function (x) { return x.propertyName === infos[0][i].propertyName; }).length > 0; }).length === infos.length) {
                    info.push(infos[0][i]);
                }
            }
            return info;
        };
    };
    CombinedObject.isPropertyDisabled = function (controls) {
        return function (name) {
            for (var i = 0; i < controls.length; i++) {
                if (controls[i]['isPropertyDisabled'](name)) {
                    return true;
                }
            }
            return false;
        };
    };
    CombinedObject.getPath = function (controls) {
        return function (name) {
            var result = '';
            for (var i = 0; i < controls.length; i++) {
                var currentPath = controls[i].getPath(name);
                if (i === 0) {
                    result = currentPath;
                }
                else if (result !== currentPath) {
                    result = '';
                    break;
                }
            }
            return result;
        };
    };
    CombinedObject.isPropertyVisible = function (controls) {
        return function (name) {
            for (var i = 0; i < controls.length; i++) {
                if (!controls[i]['isPropertyVisible'](name)) {
                    return false;
                }
            }
            return true;
        };
    };
    CombinedObject.mergeProperty = function (controls, propertyName, undoEngine, customMerge) {
        var property = controls[0][propertyName];
        var combinedObj = null;
        var subscriptions = null;
        if (controls.filter(function (x) { return !!x[propertyName]; }).length === controls.length) {
            combinedObj = customMerge && customMerge(propertyName, controls, undoEngine);
            if (!combinedObj) {
                if (ko.isObservable(property) && !property['push']) {
                    if (!controls.every(function (control) { return ko.isObservable(control[propertyName]); }))
                        return combinedObj;
                    var combinedObservable = ko.observable(controls.every(function (control) { return controls[0][propertyName].peek() === control[propertyName].peek(); }) ? controls[0][propertyName].peek() : null);
                    combinedObj = {
                        result: combinedObservable,
                        subscriptions: [combinedObservable.subscribe(function (newVal) {
                                undoEngine && undoEngine().start();
                                controls.forEach(function (control) { control[propertyName](newVal); });
                                undoEngine && undoEngine().end();
                            })]
                    };
                }
                else if (typeof property === 'object' && !$.isArray(property)) {
                    combinedObj = this._merge(controls.map(function (x) { return x[propertyName]; }), undoEngine, customMerge);
                }
            }
        }
        return combinedObj;
    };
    CombinedObject._createProperty = function (result, propertyName, propertyValue) {
        if (propertyValue) {
            if (typeof propertyValue === 'object' && _utils_1.isEmptyObject(propertyValue))
                return;
            result[propertyName] = propertyValue;
        }
    };
    CombinedObject._merge = function (controls, undoEngine, customMerge, ignoreProperties) {
        var _this = this;
        var result = {};
        var subscriptions = [];
        ['getInfo', 'isPropertyVisible', 'isPropertyDisabled', 'getPath'].forEach(function (propertyName) {
            if (controls[0][propertyName])
                _this._createProperty(result, propertyName, _this[propertyName](controls));
        });
        if (ignoreProperties) {
            var oldPropertyDisabled = result['isPropertyDisabled'];
            result['isPropertyDisabled'] = function (name) {
                return (oldPropertyDisabled && oldPropertyDisabled(name)) || ignoreProperties.indexOf(name) !== -1;
            };
        }
        if (result && result['getInfo']) {
            result['getInfo']().map(function (x) { return x.propertyName; }).forEach(function (propertyName) {
                var combinedObj = _this.mergeProperty(controls, propertyName, undoEngine, customMerge);
                if (combinedObj) {
                    subscriptions = [].concat.apply(subscriptions, combinedObj.subscriptions);
                    _this._createProperty(result, propertyName, combinedObj.result);
                }
            });
        }
        else {
            Object.keys(controls[0]).forEach(function (propertyName) {
                var combinedObj = _this.mergeProperty(controls, propertyName, undoEngine, customMerge);
                if (combinedObj) {
                    subscriptions = [].concat.apply(subscriptions, combinedObj.subscriptions);
                    _this._createProperty(result, propertyName, combinedObj.result);
                }
            });
        }
        return { result: result, subscriptions: subscriptions };
    };
    CombinedObject.mergeControls = function (controls, undoEngine, customMerge, ignoreProperties) {
        var combinedObj = this._merge(controls, undoEngine, customMerge, ignoreProperties);
        return {
            result: _utils_1.extend(combinedObj.result, { controlType: 'multiselect', displayName: ko.observable('') }),
            subscriptions: combinedObj.subscriptions
        };
    };
    CombinedObject.getEditableObject = function (selectionProvider, undoEngine, customMerge) {
        var _this = this;
        var editableObject = ko.observable(null);
        var subscriptions = [];
        var subscription = selectionProvider.focused.subscribe(function (newVal) {
            editableObject(newVal && newVal.getControlModel());
        });
        if (selectionProvider._disposables) {
            selectionProvider._disposables.push(subscription);
            selectionProvider._disposables.push({
                dispose: function () {
                    subscriptions.forEach(function (x) { return x.dispose(); });
                    subscriptions.splice(0);
                }
            });
        }
        return ko.pureComputed({
            read: function () {
                subscriptions.forEach(function (x) { return x.dispose(); });
                subscriptions = [];
                if (selectionProvider.selectedItems.length > 1) {
                    var combinedObj = _this.mergeControls(selectionProvider.selectedItems.map(function (item) { return item.getControlModel(); }), undoEngine, customMerge, selectionProvider.ignoreMultiSelectProperties);
                    subscriptions = combinedObj.subscriptions;
                    return combinedObj.result;
                }
                else {
                    return editableObject();
                }
            },
            write: function (val) {
                if (val && val.surface) {
                    selectionProvider.focused(val.surface);
                }
                else {
                    selectionProvider.updateSelection(null);
                    editableObject(val);
                }
            }
        }).extend({ deferred: true });
    };
    return CombinedObject;
}());
exports.CombinedObject = CombinedObject;
