// @devexpress/analytics-core/widgets/treelist/_treelistController.js
/**
* DevExpress Analytics (widgets\treelist\_treelistController.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TreeListController = (function () {
    function TreeListController() {
        this.selectedItem = null;
    }
    TreeListController.prototype.dispose = function () {
        this.selectedItem = null;
    };
    TreeListController.prototype.itemsFilter = function (item) {
        return true;
    };
    TreeListController.prototype.hasItems = function (item) {
        return item.specifics !== 'none' && (item.specifics === 'List' || item.specifics === 'ListSource' || item.isList === true);
    };
    TreeListController.prototype.canSelect = function (value) {
        return !value.hasItems;
    };
    TreeListController.prototype.select = function (value) {
        if (this.canSelect(value)) {
            this.selectedItem && this.selectedItem.isSelected(false);
            this.selectedItem = value;
            value.isSelected(true);
        }
    };
    return TreeListController;
}());
exports.TreeListController = TreeListController;
