// @devexpress/analytics-core/diagram/metadata.js
/**
* DevExpress Analytics (diagram\metadata.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var editorsInfo_1 = require("../property-grid/widgets/editorsInfo");
var _validation_1 = require("../core/internal/_validation");
var size_1 = require("../core/elements/size");
var point_1 = require("../core/elements/point");
exports.name = { propertyName: 'name', modelName: '@Name', displayName: 'Name', editor: editorsInfo_1.editorTemplates.getEditor('text'), validationRules: _validation_1.nameValidationRules };
exports.text = { propertyName: 'text', modelName: '@Text', displayName: 'Text', editor: editorsInfo_1.editorTemplates.getEditor('text') };
exports.size = { propertyName: 'size', modelName: '@Size', defaultVal: '100,50', from: size_1.Size.fromString, displayName: 'Size', editor: editorsInfo_1.editorTemplates.getEditor('objecteditor') };
exports.location = { propertyName: 'location', modelName: '@Location', from: point_1.Point.fromString, displayName: 'Location', editor: editorsInfo_1.editorTemplates.getEditor('objecteditor') };
exports.sizeLocation = [exports.size, exports.location];
exports.unknownSerializationsInfo = [exports.name].concat(exports.sizeLocation);
