// devextreme/cjs/ui/form/constants.js
/**
 * DevExtreme (cjs/ui/form/constants.js)
 * Version: 21.2.7
 * Build date: Mon Apr 11 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.WIDGET_CLASS = exports.SINGLE_COLUMN_ITEM_CONTENT = exports.SIMPLE_ITEM_TYPE = exports.ROOT_SIMPLE_ITEM_CLASS = exports.LAYOUT_MANAGER_ONE_COLUMN = exports.GROUP_COL_COUNT_CLASS = exports.GROUP_COL_COUNT_ATTR = exports.FORM_VALIDATION_SUMMARY = exports.FORM_UNDERLINED_CLASS = exports.FORM_LAYOUT_MANAGER_CLASS = exports.FORM_GROUP_WITH_CAPTION_CLASS = exports.FORM_GROUP_CONTENT_CLASS = exports.FORM_GROUP_CLASS = exports.FORM_GROUP_CAPTION_CLASS = exports.FORM_FIELD_ITEM_COL_CLASS = exports.FORM_CLASS = exports.FIELD_ITEM_TAB_CLASS = exports.FIELD_ITEM_LABEL_CONTENT_CLASS = exports.FIELD_ITEM_LABEL_CLASS = exports.FIELD_ITEM_CONTENT_HAS_TABS_CLASS = exports.FIELD_ITEM_CONTENT_HAS_GROUP_CLASS = exports.FIELD_ITEM_CONTENT_CLASS = exports.FIELD_ITEM_CLASS = void 0;
var WIDGET_CLASS = "dx-widget";
exports.WIDGET_CLASS = WIDGET_CLASS;
var FORM_CLASS = "dx-form";
exports.FORM_CLASS = FORM_CLASS;
var FORM_GROUP_CLASS = "dx-form-group";
exports.FORM_GROUP_CLASS = FORM_GROUP_CLASS;
var FORM_GROUP_CAPTION_CLASS = "dx-form-group-caption";
exports.FORM_GROUP_CAPTION_CLASS = FORM_GROUP_CAPTION_CLASS;
var FORM_FIELD_ITEM_COL_CLASS = "dx-col-";
exports.FORM_FIELD_ITEM_COL_CLASS = FORM_FIELD_ITEM_COL_CLASS;
var FIELD_ITEM_CLASS = "dx-field-item";
exports.FIELD_ITEM_CLASS = FIELD_ITEM_CLASS;
var LAYOUT_MANAGER_ONE_COLUMN = "dx-layout-manager-one-col";
exports.LAYOUT_MANAGER_ONE_COLUMN = LAYOUT_MANAGER_ONE_COLUMN;
var FIELD_ITEM_LABEL_CONTENT_CLASS = "dx-field-item-label-content";
exports.FIELD_ITEM_LABEL_CONTENT_CLASS = FIELD_ITEM_LABEL_CONTENT_CLASS;
var FORM_LAYOUT_MANAGER_CLASS = "dx-layout-manager";
exports.FORM_LAYOUT_MANAGER_CLASS = FORM_LAYOUT_MANAGER_CLASS;
var FIELD_ITEM_LABEL_CLASS = "dx-field-item-label";
exports.FIELD_ITEM_LABEL_CLASS = FIELD_ITEM_LABEL_CLASS;
var FIELD_ITEM_CONTENT_CLASS = "dx-field-item-content";
exports.FIELD_ITEM_CONTENT_CLASS = FIELD_ITEM_CONTENT_CLASS;
var SINGLE_COLUMN_ITEM_CONTENT = "dx-single-column-item-content";
exports.SINGLE_COLUMN_ITEM_CONTENT = SINGLE_COLUMN_ITEM_CONTENT;
var ROOT_SIMPLE_ITEM_CLASS = "dx-root-simple-item";
exports.ROOT_SIMPLE_ITEM_CLASS = ROOT_SIMPLE_ITEM_CLASS;
var FORM_GROUP_CONTENT_CLASS = "dx-form-group-content";
exports.FORM_GROUP_CONTENT_CLASS = FORM_GROUP_CONTENT_CLASS;
var FIELD_ITEM_CONTENT_HAS_GROUP_CLASS = "dx-field-item-has-group";
exports.FIELD_ITEM_CONTENT_HAS_GROUP_CLASS = FIELD_ITEM_CONTENT_HAS_GROUP_CLASS;
var FIELD_ITEM_CONTENT_HAS_TABS_CLASS = "dx-field-item-has-tabs";
exports.FIELD_ITEM_CONTENT_HAS_TABS_CLASS = FIELD_ITEM_CONTENT_HAS_TABS_CLASS;
var FORM_GROUP_WITH_CAPTION_CLASS = "dx-form-group-with-caption";
exports.FORM_GROUP_WITH_CAPTION_CLASS = FORM_GROUP_WITH_CAPTION_CLASS;
var FIELD_ITEM_TAB_CLASS = "dx-field-item-tab";
exports.FIELD_ITEM_TAB_CLASS = FIELD_ITEM_TAB_CLASS;
var GROUP_COL_COUNT_CLASS = "dx-group-colcount-";
exports.GROUP_COL_COUNT_CLASS = GROUP_COL_COUNT_CLASS;
var GROUP_COL_COUNT_ATTR = "group-col-count";
exports.GROUP_COL_COUNT_ATTR = GROUP_COL_COUNT_ATTR;
var FORM_VALIDATION_SUMMARY = "dx-form-validation-summary";
exports.FORM_VALIDATION_SUMMARY = FORM_VALIDATION_SUMMARY;
var FORM_UNDERLINED_CLASS = "dx-form-styling-mode-underlined";
exports.FORM_UNDERLINED_CLASS = FORM_UNDERLINED_CLASS;
var SIMPLE_ITEM_TYPE = "simple";
exports.SIMPLE_ITEM_TYPE = SIMPLE_ITEM_TYPE;
