// @devexpress/analytics-core/query-builder/binding/jsQueryBuilder.js
/**
* DevExpress Analytics (query-builder\binding\jsQueryBuilder.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var localization_utils_1 = require("../../property-grid/localization/localization_utils");
var _infoMessageHelpers_1 = require("../../core/utils/_infoMessageHelpers");
var _requestwrapper_1 = require("../utils/_requestwrapper");
var JSQueryBuilder = (function () {
    function JSQueryBuilder(_queryBuilderModel) {
        this._queryBuilderModel = _queryBuilderModel;
    }
    Object.defineProperty(JSQueryBuilder.prototype, "queryBuilderModel", {
        get: function () {
            return this._queryBuilderModel();
        },
        set: function (newVal) {
            this._queryBuilderModel(newVal);
        },
        enumerable: true,
        configurable: true
    });
    JSQueryBuilder.prototype.UpdateLocalization = function (localization) {
        localization_utils_1.updateLocalization(localization);
    };
    JSQueryBuilder.prototype.GetQueryBuilderModel = function () {
        return this.queryBuilderModel;
    };
    JSQueryBuilder.prototype.GetJsonQueryModel = function () {
        return { 'Query': this.queryBuilderModel.model().serialize() };
    };
    JSQueryBuilder.prototype.GetSaveQueryModel = function () {
        return {
            queryLayout: JSON.stringify(this.GetJsonQueryModel()),
            connection: this.SerializeDataConnection()
        };
    };
    JSQueryBuilder.prototype.SerializeDataConnection = function () {
        var connection = this.queryBuilderModel.model().dbSchemaProvider.connection;
        return _requestwrapper_1.serializeDataConnection(connection);
    };
    JSQueryBuilder.prototype.AdjustControlCore = function () {
        this.queryBuilderModel && this.queryBuilderModel.updateSurfaceSize();
    };
    JSQueryBuilder.prototype.Save = function () {
        this.queryBuilderModel && this.queryBuilderModel.model().onSave();
    };
    JSQueryBuilder.prototype.ShowPreview = function () {
        this.queryBuilderModel && this.queryBuilderModel.showPreview();
    };
    JSQueryBuilder.prototype.IsQueryValid = function () {
        return this.queryBuilderModel && this.queryBuilderModel.model().isValid();
    };
    JSQueryBuilder.prototype.OnCallback = function (result) {
        if (result.queryValidationError) {
            _infoMessageHelpers_1.NotifyAboutWarning(result.queryValidationError, true);
        }
    };
    return JSQueryBuilder;
}());
exports.JSQueryBuilder = JSQueryBuilder;
