// @devexpress/analytics-core/core/snapLines/_snapLinesCollector.js
/**
* DevExpress Analytics (core\snapLines\_snapLinesCollector.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _arrayutils_1 = require("../utils/_arrayutils");
var SnapLinesCollector = (function () {
    function SnapLinesCollector() {
        this._verticalSnapLines = [];
        this._horizontalSnapLines = [];
        this._snapTargetToIgnore = null;
    }
    SnapLinesCollector.prototype._appendSnapLine = function (position, limitInf, limitSup, snapLines) {
        var line = {
            position: position,
            limitInf: limitInf,
            limSup: limitSup
        };
        var index = _arrayutils_1.binaryIndexOf(snapLines, line, function (a, b) { return a.position - b.position; });
        if (index > -1) {
            snapLines[index].limitInf = Math.min(snapLines[index].limitInf, limitInf);
            snapLines[index].limSup = Math.max(snapLines[index].limSup, limitSup);
        }
        else {
            snapLines.splice(~index, 0, line);
        }
    };
    SnapLinesCollector.prototype._collectSnaplines = function (parent, parentAbsoluteProsition) {
        var _this = this;
        this._enumerateCollection(parent, parentAbsoluteProsition, function (item, itemAbsoluteRect) {
            if (item !== _this._snapTargetToIgnore) {
                _this._appendSnapLine(itemAbsoluteRect.left, itemAbsoluteRect.top, itemAbsoluteRect.bottom, _this._verticalSnapLines);
                _this._appendSnapLine(itemAbsoluteRect.right, itemAbsoluteRect.top, itemAbsoluteRect.bottom, _this._verticalSnapLines);
                _this._appendSnapLine(itemAbsoluteRect.top, itemAbsoluteRect.left, itemAbsoluteRect.right, _this._horizontalSnapLines);
                _this._appendSnapLine(itemAbsoluteRect.bottom, itemAbsoluteRect.left, itemAbsoluteRect.right, _this._horizontalSnapLines);
                _this._collectSnaplines(item, itemAbsoluteRect);
            }
        });
    };
    SnapLinesCollector.prototype._getCollection = function (parent) {
        return parent['controls'] && parent['controls']();
    };
    SnapLinesCollector.prototype._enumerateCollection = function (parent, parentAbsoluteProsition, callback) {
        var collection = this._getCollection(parent);
        if (!collection)
            return;
        for (var i = 0; i < collection.length; i++) {
            var itemRect = collection[i].rect && collection[i].rect();
            if (itemRect) {
                callback(collection[i], {
                    top: itemRect.top + parentAbsoluteProsition.top,
                    bottom: itemRect.bottom + parentAbsoluteProsition.top,
                    left: itemRect.left + parentAbsoluteProsition.left,
                    right: itemRect.right + parentAbsoluteProsition.left
                });
            }
        }
    };
    SnapLinesCollector.prototype.collectSnaplines = function (root, snapTargetToIgnore) {
        this._snapTargetToIgnore = snapTargetToIgnore;
        this._verticalSnapLines.splice(0);
        this._horizontalSnapLines.splice(0);
        this._collectSnaplines(root, { top: 0, left: 0 });
        return {
            vertical: this._verticalSnapLines,
            horizontal: this._horizontalSnapLines
        };
    };
    return SnapLinesCollector;
}());
exports.SnapLinesCollector = SnapLinesCollector;
