// @devexpress/analytics-core/property-grid/widgets/internal/_internal.js
/**
* DevExpress Analytics (property-grid\widgets\internal\_internal.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _internal_1 = require("../../../serializer/_internal");
var localization_utils_1 = require("../../localization/localization_utils");
exports.propertiesGridEditorsPaddingLeft = _internal_1.createGlobalModuleVariableFunc(19);
exports.defaultFontSerialization = _internal_1.createGlobalModuleVariableFunc('Times New Roman, 9pt');
function RegexGuid(guid) {
    return (/^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/.test(guid)
        || /^\{[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\}$/.test(guid)
        || /^\([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\)$/.test(guid)
        || /^[0-9a-fA-F]{32}$/.test(guid));
}
function validateGuid(guid) {
    return guid && RegexGuid(guid);
}
exports.validateGuid = validateGuid;
function validateNullableGuid(guid) {
    return !guid || RegexGuid(guid);
}
exports.validateNullableGuid = validateNullableGuid;
var guidValidationMessage = function () { return localization_utils_1.getLocalization('Guid is required and should have a valid format.', 'AnalyticsCoreStringId.GuidIsRequired_Error'); };
exports.guidValidationRules = [{
        type: 'custom',
        validationCallback: function (options) { return validateNullableGuid(options.value); },
        get message() {
            return guidValidationMessage();
        }
    }];
exports.guidRequiredValidationRules = [{
        type: 'required',
        get message() {
            return guidValidationMessage();
        }
    }];
var requiredValidationRule = {
    type: 'required',
    get message() {
        return localization_utils_1.getLocalization('The value cannot be empty', 'AnalyticsCoreStringId.ParametersPanel_DateTimeValueValidationError');
    }
};
exports.requiredValidationRules = [requiredValidationRule];
