// devexpress-reporting/designer/controls/crossTab/enums.js
/**
* DevExpress HTML/JS Reporting (designer\controls\crossTab\enums.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var CornerHeaderDisplayMode;
(function (CornerHeaderDisplayMode) {
    CornerHeaderDisplayMode[CornerHeaderDisplayMode["None"] = 0] = "None";
    CornerHeaderDisplayMode[CornerHeaderDisplayMode["RowFieldNames"] = 1] = "RowFieldNames";
    CornerHeaderDisplayMode[CornerHeaderDisplayMode["ColumnFieldNames"] = 2] = "ColumnFieldNames";
})(CornerHeaderDisplayMode = exports.CornerHeaderDisplayMode || (exports.CornerHeaderDisplayMode = {}));
var CellKind;
(function (CellKind) {
    CellKind[CellKind["None"] = 0] = "None";
    CellKind[CellKind["Corner"] = 1] = "Corner";
    CellKind[CellKind["RowHeader"] = 2] = "RowHeader";
    CellKind[CellKind["RowTotalHeader"] = 3] = "RowTotalHeader";
    CellKind[CellKind["RowTotal"] = 4] = "RowTotal";
    CellKind[CellKind["ColumnHeader"] = 5] = "ColumnHeader";
    CellKind[CellKind["ColumnTotalHeader"] = 6] = "ColumnTotalHeader";
    CellKind[CellKind["ColumnTotal"] = 7] = "ColumnTotal";
    CellKind[CellKind["Data"] = 8] = "Data";
    CellKind[CellKind["DataHeader"] = 9] = "DataHeader";
    CellKind[CellKind["GrandTotal"] = 10] = "GrandTotal";
    CellKind[CellKind["EmptyHeader"] = 11] = "EmptyHeader";
    CellKind[CellKind["Empty"] = 12] = "Empty";
})(CellKind = exports.CellKind || (exports.CellKind = {}));
var TotalHeaderPosition;
(function (TotalHeaderPosition) {
    TotalHeaderPosition[TotalHeaderPosition["Inner"] = 0] = "Inner";
    TotalHeaderPosition[TotalHeaderPosition["Outer"] = 1] = "Outer";
})(TotalHeaderPosition = exports.TotalHeaderPosition || (exports.TotalHeaderPosition = {}));
var TotalsPosition;
(function (TotalsPosition) {
    TotalsPosition[TotalsPosition["AfterData"] = 0] = "AfterData";
    TotalsPosition[TotalsPosition["BeforeData"] = 1] = "BeforeData";
})(TotalsPosition = exports.TotalsPosition || (exports.TotalsPosition = {}));
var DataFieldLayout;
(function (DataFieldLayout) {
    DataFieldLayout[DataFieldLayout["InRow"] = 0] = "InRow";
    DataFieldLayout[DataFieldLayout["InColumn"] = 1] = "InColumn";
})(DataFieldLayout = exports.DataFieldLayout || (exports.DataFieldLayout = {}));
