// @devexpress/analytics-core/query-builder/dataSource/object/objectDataSource.js
/**
* DevExpress Analytics (query-builder\dataSource\object\objectDataSource.js)
* Version:  21.2.7
* Build date: Apr 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var utils_1 = require("../../../serializer/utils");
var _utils_1 = require("../../../undo-engine/_utils");
var ObjectDataSource = (function (_super) {
    __extends(ObjectDataSource, _super);
    function ObjectDataSource() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = ko.observable();
        return _this;
    }
    ObjectDataSource.prototype.setState = function (state) {
        this.ctor = state.ctor;
        this.dataMember = state.dataMember;
        this.selectedType = state.selectedType;
        this.id = _utils_1.guid().replace(/-/g, '');
        this.name(state.dataSourceName);
    };
    return ObjectDataSource;
}(utils_1.Disposable));
exports.ObjectDataSource = ObjectDataSource;
